test/replication-luatest/election_strict_fencing_test.lua

```diff
@@ -0,0 +1,108 @@
+local luatest = require('luatest')
+local cluster = require('test.luatest_helpers.cluster')
+local server = require('test.luatest_helpers.server')
+local g = luatest.group('fencing_mode', {
+    {election_fencing_mode = 'soft'}, {election_fencing_mode = 'strict'}})
+
+local function promote(server)
+    luatest.helpers.retrying({}, function()
+        luatest.assert(server:exec(function()
+            pcall(box.ctl.promote)
+            if box.info.election.state ~= 'leader' then
+                return false
+            end
+            if box.info.synchro.queue.owner ~= box.info.id then
+                return false
+            end
+            return true
+        end))
+    end)
+    server:exec(function() box.ctl.wait_rw() end)
+end
+
+local function wait_sync(leader, servers)
+    local vclock = leader:get_vclock()
+    vclock[0] = nil
+    for _, server in ipairs(servers) do
+        server:wait_vclock(vclock)
+    end
+end
+
+local function box_cfg_update(servers, cfg)
+    for _, server in ipairs(servers) do
+        server:exec(function(cfg) box.cfg(cfg) end, {cfg})
+    end
+end
+
+g.before_all(function(g)
+    g.box_cfg = {
+        election_mode = 'manual',
+        election_timeout = 0.1,
+        replication = {
+            server.build_instance_uri('server_1'),
+            server.build_instance_uri('server_2'),
+        },
+        replication_synchro_quorum = 2,
+        replication_synchro_timeout = 0.1,
+        replication_timeout = 0.1,
+    }
+
+    g.cluster = cluster:new({})
+    g.server_1 = g.cluster:build_and_add_server(
+        {alias = 'server_1', box_cfg = g.box_cfg})
+
+    g.box_cfg.read_only = true
+    g.server_2 = g.cluster:build_and_add_server(
+        {alias = 'server_2', box_cfg = g.box_cfg})
+
+    g.cluster:start()
+    g.cluster:wait_fullmesh()
+    promote(g.server_1)
+    wait_sync(g.server_1, g.cluster.servers)
+end)
+
+g.after_all(function(g)
+    g.server_2:stop()
+    g.server_1:stop()
+end)
+
+g.test_fencing = function(g)
+    local timeout = 0.5
+    box_cfg_update(g.cluster.servers, {
+        election_fencing_mode = g.params.election_fencing_mode,
+        replication_timeout = timeout,
+    })
+    box_cfg_update({g.server_2}, {
+        election_mode = 'candidate',
+        replication_synchro_quorum = 1,
+    })
+    promote(g.server_1)
+    wait_sync(g.server_1, g.cluster.servers)
+
+    -- Wait for first heartbeat
+    require('fiber').sleep(timeout)
+
+    local old_term = g.server_1:election_term()
+
+    g.server_1:exec(function()
+        box.error.injection.set('ERRINJ_COIO_WRITE_DROP', true)
```

Comment:
If I understand correctly, you use this injection to prevent server's relay from sending anything.
Couldn't you use `ERRINJ_RELAY_SEND_DELAY` instead?

Reply:
No, the behaviour is different: when ERRINJ_RELAY_SEND_DELAY is on relay_send yields and Tarantool never thinks that it sent the message, with ERRINJ_COIO_WRITE_DROP Tarantool thinks that packet was sent correctly. I wanted to simulate network problems outside Tarantool.

Reply:
Yes, indeed. If you used ERRINJ_RELAY_SEND_DELAY, relay would never update `relay->last_row_time`.
But does this matter for your test?
It still wouldn't send anything to the replica. And relay's disconnect timeout is not connected to `last_row_time` anyhow.

Reply:
Sorry, I kinda forgot that the problem was that with ERRINJ_RELAY_SEND_DELAY server_1 still though server_2 was connected and never fenced, just box.info.downstream.idle was growing.

Reply:
Upon further investigation turned out that enabling ERRINJ_RELAY_SEND_DELAY makes relay_subscribe_f stuck on sending next heartbeat. Even though relay_reader_f gets timedout, it's never joined by relay_subscribe_f (which created it) and tarantool doesn't notice connection failure. IMO this isn't a real bug, but rather ERRINJ_RELAY_SEND_DELAY thing.

Reply:
Yes, indeed. Thanks for digging into this!